File: src/main/java/editor/database/symbol/PhyrexianHybridSymbol.java
package editor.database.symbol;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

import editor.database.attributes.ManaType;

class PhyrexianHybridSymbol extends ManaSymbol {
    public static final Map<ManaType, Map<ManaType, PhyrexianHybridSymbol>> SYMBOLS = Collections.unmodifiableMap(Arrays.stream(ManaType.colors()).collect(
        Collectors.toMap(
            Function.identity(),
            (m) -> Arrays.stream(ManaType.colors()).filter((n) -> n != m).collect(Collectors.toMap(Function.identity(), (n) -> new PhyrexianHybridSymbol(m, n)))
        )
    ));

    public static PhyrexianHybridSymbol parsePhyrexianHybridSymbol(String pair) throws IllegalArgumentException
    {
        return tryParsePhyrexianHybridSymbol(pair).orElseThrow(() -> new IllegalArgumentException('"' + pair + "\" is not a Phyrexian hybrid symbol"));
    }

    public static Optional<PhyrexianHybridSymbol> tryParsePhyrexianHybridSymbol(String pair)
    {
        var tokens = pair.split("/");
        if (tokens.length == 3 && tokens[2].equals("P"))
        {
            var c1 = ManaType.tryParseManaType(tokens[0]);
            var c2 = ManaType.tryParseManaType(tokens[1]);
            if (SYMBOLS.get(c1) != null)
                return Optional.ofNullable(SYMBOLS.get(c1).get(c2));
        }
        return Optional.empty();
    }

    private final ManaType color1;
    private final ManaType color2;

    private PhyrexianHybridSymbol(ManaType col1, ManaType col2)
    {
        super("phyrexian_" + (col1.colorOrder(col2) > 0 ? col2 : col1).toString().toLowerCase() + '_' + (col1.colorOrder(col2) > 0 ? col1 : col2).toString().toLowerCase() + "_mana.png",
                (col1.colorOrder(col2) > 0 ? col2 : col1).shorthand() + "/" + (col1.colorOrder(col2) > 0 ? col1 : col2).shorthand() + "/P",
                1);
        color1 = col1;
        color2 = col2;
    }

    @Override
    public Map<ManaType, Double> colorIntensity()
    {
        return createIntensity(new ColorIntensity(color1, 1.0/3.0), new ColorIntensity(color2, 1.0/3.0));
    }

    @Override
    public int compareTo(ManaSymbol o)
    {
        if (o instanceof PhyrexianHybridSymbol s)
            return color1.compareTo(s.color1)*10 + color2.compareTo(s.color2);
        else
            return super.compareTo(o);
    }
}

File: src/main/scala/editor/database/symbol/PhyrexianHybridSymbol.scala
package editor.database.symbol

import editor.database.attributes.ManaType

import scala.jdk.CollectionConverters._
import scala.jdk.OptionConverters._

/**
 * Companion to [[PhyrexianHybridSymbol]] that contains all possible symbols and methods for parsing them from strings.
 * @author Alec Roelke
 */
object PhyrexianHybridSymbol {
  @deprecated def SYMBOLS = ManaType.colors.map((c1) => c1 -> ManaType.colors.collect{ case (c2) if c2 != c1 => c2 -> PhyrexianHybridSymbolGenerator.values((c1, c2)) })
  @deprecated def tryParsePhyrexianHybridSymbol(s: String) = PhyrexianHybridSymbolGenerator.parse(s).toJava
  @deprecated def parsePhyrexianHybridSymbol(s: String) = {
    val colors = s.split("/").map(ManaType.parseManaType)
    PhyrexianHybridSymbolGenerator.values((colors(0), colors(1)))
  }
}

/**
 * A "Phyrexian" mana symbol that can be paid with either of two colors of mana or 2 life. Its color intensity is 1/3 for each of its two colors of mana,
 * and 0 for the rest, as it has three ways to pay for it (one of two colors or 2 life), and its text representation is the shorthand of its colors and a P
 * separated by a slash ("/").
 * 
 * @constructor create a new Phyrexian hybrid mana symbol
 * @param first first color that can be used to pay
 * @param second second color that can be used to pay
 * 
 * @author Alec Roelke
 */
class PhyrexianHybridSymbol private[symbol](val first: ManaType, val second: ManaType) extends ManaSymbol(s"phyrexian_${first.toString.toLowerCase}_${second.toString.toLowerCase}_mana.png", s"${first.shorthand.toUpper}/${second.shorthand.toUpper}/P", 1, PhyrexianHybridSymbolGenerator) {
  override def colorIntensity = ManaSymbol.createIntensity(Map(first -> 1.0/3.0, second -> 1.0/3.0))
}
